File: frontend/src/api/providers.ts
import { request } from "./http";

export type Provider = {
  id: string;
  name: string;
  protocol: string;
  base_url: string;
  capability: string;
  status: string;
  latency_ms?: number | null;
  models: string[];
  notes: string;
  key_masked: string;
  has_key: boolean;
};

export type RoleBinding = {
  id: string;
  role: string;
  label: string;
  provider_id?: string | null;
  model: string;
  /** 语音角色的 TTS 槽；ASR 仍用 provider_id / model。 */
  tts_provider_id?: string | null;
  tts_model?: string;
  temperature: number;
};

export const providerApi = {
  adminOverview: () => request<Record<string, unknown>>("/api/admin/overview"),
  providers: () => request<Provider[]>("/api/admin/providers"),
  createProvider: (body: Partial<Provider> & { api_key?: string }) =>
    request<Provider>("/api/admin/providers", { method: "POST", body: JSON.stringify(body) }),
  updateProvider: (id: string, body: Partial<Provider> & { api_key?: string }) =>
    request<Provider>(`/api/admin/providers/${id}`, { method: "PATCH", body: JSON.stringify(body) }),
  pingProvider: (id: string) => request<Provider>(`/api/admin/providers/${id}/ping`, { method: "POST" }),
  deleteProvider: (id: string) => request<{ ok: string }>(`/api/admin/providers/${id}/delete`, { method: "POST" }),
  probeModels: (body: { protocol: string; base_url: string; api_key?: string; provider_id?: string }) =>
    request<{ models: string[] }>("/api/admin/provider-models", { method: "POST", body: JSON.stringify(body) }),
  roles: () => request<RoleBinding[]>("/api/admin/roles"),
  updateRole: (role: string, body: Partial<RoleBinding>) =>
    request<RoleBinding>(`/api/admin/roles/${role}`, { method: "PATCH", body: JSON.stringify(body) }),
};

File: backend/app/schemas/providers.py
"""Request and response shapes for vendors, role bindings, and the admin overview."""

from pydantic import BaseModel, Field


class ProviderOut(BaseModel):
    id: str
    name: str
    protocol: str
    base_url: str
    capability: str
    status: str
    latency_ms: int | None = None
    models: list[str] = Field(default_factory=list)
    notes: str = ""
    key_masked: str = ""
    has_key: bool = False


class ProviderProbeIn(BaseModel):
    protocol: str = "anthropic_messages"
    base_url: str = ""
    api_key: str | None = None
    provider_id: str | None = None


class ProviderIn(BaseModel):
    name: str | None = None
    protocol: str | None = None
    base_url: str | None = None
    capability: str | None = None
    api_key: str | None = None
    models: list[str] | None = None
    notes: str | None = None


class RoleBindingOut(BaseModel):
    id: str
    role: str
    label: str
    provider_id: str | None = None
    model: str = ""
    # 仅语音角色使用：与 provider_id/model（ASR）成对的 TTS 绑定。
    tts_provider_id: str | None = None
    tts_model: str = ""
    temperature: float = 0.4


class RoleBindingIn(BaseModel):
    """Partial update. Unset fields must stay unset so a one-slot edit cannot wipe the other."""

    provider_id: str | None = None
    model: str | None = None
    tts_provider_id: str | None = None
    tts_model: str | None = None
    temperature: float | None = None


class AdminOverview(BaseModel):
    providers: list[ProviderOut]
    roles: list[RoleBindingOut] = Field(default_factory=list)
    metrics: dict[str, str]
    material_count: int = 0
    chunk_count: int = 0
    index_status: str = "就绪"
    recall_score: str = "—"
    calls_24h: int = 0
    fail_rate: str = "0%"
    p95_ms: int = 0
